test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,1159 @@
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x cursor(1, 1)[-1].'x'")
+
+      feed(":%s/tw/tox/g<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
+    end
+  end)
+
+  it('work when a mapping disables incsub', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x execute('set incsubstitute=')[-1]")
+
+      feed(":%s/tw/toxa/g<enter>")
+      expect(string.gsub(default_text, "tw", "toa"))
+    end
+  end)
+
+  it('work with a complex mapping', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      source([[cnoremap x <C-\>eextend(g:, {'fo': getcmdline()})
+      \.fo<CR><C-c>:new<CR>:bw!<CR>:<C-r>=remove(g:, 'fo')<CR>x]])
+
+      feed(":%s/tw/tox")
+      feed("/<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
+    end
+  end)
+
+end)
+
+-- *** AUTOCMDS
+-- Right now, none fire when opening or closing the incsub buffer
+--
+-- If that behavior is changed, add tests (suggestions by ZyX-l):
+-- * If any are run when window is opened, then tests which check what if these
+--   autocommands close the window.
+-- * If any are run when window is opened, then
+--   tests which check what if these autocommands wipe out buffer without closing
+--   the window.
+-- * If any are run when window is opened, then tests which check
+--   what if these autocommands switch buffer without closing the window or
+--   wiping buffer out.
+-- * If any are run when window is opened, then tests which
+--   check what if these autocommands wipe out buffer where substitution is
+--   being made.
+-- * If any are run when window is opened, then tests which check what
+--   if these autocommands close window where substitution is being made (but
+--   not window with temporary view)
+
+describe('Incsubstitute: autocommands', function()
+  before_each(clear)
+
+  -- keys are events to be tested
+  -- values are arrays like { open = 1, close = 2} which would mean
+  -- the event is expected to fire for buffer number 1 when the incsub 
+  -- window is opend, and for buffer number 2 when the incsub window is
+  -- closed
+  local eventsExpected = {
+    BufAdd = {},
+    BufDelete = {open = 2},
+    BufEnter = {open = 1},
+    BufFilePost = {},
+    BufFilePre = {},
+    BufHidden = {open = 2},
+    BufLeave = {open = 2},
+    BufNew = {},
+    BufNewFile = {},
+    BufRead = {},
+    BufReadCmd = {},
+    BufReadPre = {},
+    BufUnload = {open = 2},
+    BufWinEnter = {},
+    BufWinLeave = {open = 2},
+    BufWipeout = {open = 2},
+    BufWrite = {},
+    BufWriteCmd = {},
+    BufWritePost = {},
+    Syntax = {},
+    FileType = {},
+    WinEnter = {open = 1},
+    WinLeave = {open = 2},
+    CmdwinEnter = {},
+    CmdwinLeave = {open = 1},
+  }
+
+  local function register_autocmd(event)
+    execute("let g:" .. event .. "_fired=0")
+    execute("autocmd " .. event .. " * let g:" .. event .. "_fired=expand('<abuf>')")
```

Comment:
What if event was fired more then once? I think this should be set to a list in first execute and use `call add(…)` here.